File: snippets/python/cht.py
from collections import deque

class CHT:
    """
    Convex Hull Trick for DP optimization
    Maintains lines y = mx + b, answers min queries
    
    Requirements:
    - Lines added in decreasing order of slope (m)
    - Queries in increasing order of x
    
    Time: O(1) amortized per operation
    
    Use case: dp[i] = min(dp[j] + cost(j,i)) where cost(j,i) = m[j] * x[i] + b[j]
    """
    def __init__(self):
        self.lines = deque()  # (m, b)
    
    def _bad(self, l1, l2, l3) -> bool:
        """Check if l2 is useless (dominated by l1 and l3)"""
        # (b1-b2)/(m2-m1) >= (b2-b3)/(m3-m2)
        # Cross multiply to avoid division
        return (l1[1] - l2[1]) * (l3[0] - l2[0]) >= (l2[1] - l3[1]) * (l2[0] - l1[0])
    
    def add(self, m: int, b: int) -> None:
        """Add line y = mx + b (must add in decreasing order of m)"""
        line = (m, b)
        while len(self.lines) >= 2 and self._bad(self.lines[-2], self.lines[-1], line):
            self.lines.pop()
        self.lines.append(line)
    
    def query(self, x: int) -> int:
        """Query minimum y at x (must query in increasing order of x)"""
        while len(self.lines) >= 2:
            m1, b1 = self.lines[0]
            m2, b2 = self.lines[1]
            if m1 * x + b1 >= m2 * x + b2:
                self.lines.popleft()
            else:
                break
        m, b = self.lines[0]
        return m * x + b

File: snippets/rust/cht.rs
// Convex Hull Trick (monotonic) for min queries.
// Add lines `y = m*x + b` in order of decreasing slope `m`; query `x` in increasing order.
struct CHT {
    lines: std::collections::VecDeque<(i64, i64)>,
}
impl CHT {
    fn new() -> Self { Self { lines: std::collections::VecDeque::new() } }
    fn bad(l1: (i64, i64), l2: (i64, i64), l3: (i64, i64)) -> bool {
        (l1.1 - l2.1) as i128 * (l3.0 - l2.0) as i128
            >= (l2.1 - l3.1) as i128 * (l2.0 - l1.0) as i128
    }
    fn add(&mut self, m: i64, b: i64) {
        let line = (m, b);
        while self.lines.len() >= 2 {
            let l1 = self.lines[self.lines.len() - 2];
            let l2 = *self.lines.back().unwrap();
            if Self::bad(l1, l2, line) { self.lines.pop_back(); } else { break; }
        }
        self.lines.push_back(line);
    }
    fn query(&mut self, x: i64) -> i64 {
        while self.lines.len() >= 2 {
            let a = self.lines[0]; let b = self.lines[1];
            if a.0 * x + a.1 >= b.0 * x + b.1 { self.lines.pop_front(); } else { break; }
        }
        self.lines[0].0 * x + self.lines[0].1
    }
}
